juju/juju
# feat(dqlite): double write Dead machine life with EnsureDeadMachine backed by dqlite

This is a companion PR for #17627. After wiring up the dqlite backed machine watcher into the provisioner worker, we noticed that the provisioner no longer gets any changes when a machine goes to `Dead` status (for machine removal). So we need to double-write the dead so when the provisioner is updated with the new watcher it can get the changes for Dead machines.

So this:

- adds a new `SetMachineLife` service on the machine domain.
- adds a new `EnsureDeadMachine` service that uses the `SetMachineLife` to set a machine's life to `life.Dead`.
- wires up the machine service into the `common.DeadEnsurer` to use the `EnsureDeadMachine` to double write the life if the entity that we're ensuring dead is a machine.
- also wires up the machine service through the `machiner`, `provisioner`, `storageprovisioner`, and `uniter` agent APIs on the apiserver that all uses the `DeadEnsurer`. 


## Checklist




## QA steps

> [!WARNING]
> I'm not entirely sure how to QA this. We already double write the machine at creation. But we don't have anything that uses life info from dqlite just yet. I think the best (and maybe the only) way to QA this change is to manually check the machine life on dqlite after calling `remove-machine` to see if it's set to Dead. However, I've been unsuccessful with the dqlite repl so far.

## Links



**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/

**Jira card:** JUJU-6341
